File: lintlib/src/main/java/com/xgimi/lintlib/detector/XGimiLogDetector.java
package com.xgimi.lintlib.detector;

import com.android.tools.lint.client.api.UElementHandler;
import com.android.tools.lint.detector.api.Category;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Implementation;
import com.android.tools.lint.detector.api.Issue;
import com.android.tools.lint.detector.api.JavaContext;
import com.android.tools.lint.detector.api.Scope;
import com.android.tools.lint.detector.api.Severity;
import com.xgimi.lintlib.util.ConfigurationTitle;

import org.jetbrains.annotations.Nullable;
import org.jetbrains.uast.UElement;
import org.jetbrains.uast.ULiteralExpression;
import org.jetbrains.uast.UastLiteralUtils;

import java.util.Collections;
import java.util.List;

/**
 * @author chongyang.zhang
 * @date 2018/7/28
 * maintainer chongyang.zhang
 * copyright 2018 www.xgimi.com Inc. All rights reserved.
 * desc:
 */
public class XGimiLogDetector extends Detector implements Detector.UastScanner {

    public static final Issue ISSUE = Issue.create(ConfigurationTitle.addTitle("Log"),
            "1231",
            "adsdf",
            Category.CORRECTNESS, 6, Severity.WARNING,
            new Implementation(XGimiLogDetector.class, Scope.JAVA_FILE_SCOPE));

    @Override
    public List<Class<? extends UElement>> getApplicableUastTypes() {
        return Collections.singletonList(ULiteralExpression.class);
    }

    @Nullable
    @Override
    public UElementHandler createUastHandler(JavaContext context) {
        return new UElementHandler() {
            @Override
            public void visitLiteralExpression(ULiteralExpression expression) {
                String string = UastLiteralUtils.getValueIfStringLiteral(expression);
                if (string == null) {
                    return;
                }
                if (string.contains("Log") && string.matches(".*\\bLog\\b.*")) {
                    context.report(ISSUE, expression, context.getLocation(expression),
                            "error");
                }
            }
        };
    }
}

File: lintlib/src/main/java/com/xgimi/lintlib/detector/log/XGimiLogDetector.kt
package com.xgimi.lintlib.detector.log

import com.android.tools.lint.detector.api.*
import com.intellij.psi.PsiMethod
import com.xgimi.lintlib.util.ConfigurationTitle
import org.jetbrains.uast.UCallExpression
import java.util.*

/**
 * @author chongyang.zhang
 * @date 2018/7/30
 * maintainer chongyang.zhang
 * copyright 2018 www.xgimi.com Inc. All rights reserved.
 * desc:
 */
class XGimiLogDetector : Detector(), SourceCodeScanner {
    companion object Issues {
        private val IMPLEMENTATION = Implementation(
                XGimiLogDetector::class.java, Scope.JAVA_FILE_SCOPE)
        
        @JvmField
        val ISSUE = Issue.create(
                LogConstant.ID + "Log",
                LogConstant.DESCRIPTION,
                LogConstant.EXPLANATION,
                LogConstant.CATEGORY,
                LogConstant.LEVEL,
                LogConstant.SEVERITY,
                IMPLEMENTATION)
        
        private const val IS_LOGGABLE = "isLoggable"
        private const val LOG_CLS = "android.util.Log"
        private const val PRINTLN = "println"
        
        val logFix: LintFix = LintFix
                .create()
                .name(ConfigurationTitle.addTitle("使用${LogConstant.LOG}"))
                .replace()
                .all()
                .pattern("Log")
                .with(LogConstant.LOG)
                .build()
        
    }
    
    override fun getApplicableMethodNames(): List<String>? =
            Arrays.asList(
                    "d",
                    "e",
                    "i",
                    "v",
                    "w",
                    PRINTLN,
                    IS_LOGGABLE)
    
    override fun visitMethod(context: JavaContext, node: UCallExpression,
                             method: PsiMethod) {
        val evaluator = context.evaluator
        if (!evaluator.isMemberInClass(method, LOG_CLS)) {
            return
        }
        if (context.isEnabled(ISSUE)) {
            context.report(ISSUE,
                    node,
                    context.getLocation(node),
                    LogConstant.MESSAGE,
                    logFix)
        }
    }
    
}